[CCAP-570] ask families for employment start date

enyia21 opened the pull request:
#### 🔗 Jira ticket
CCAP-570

#### ✍️ Description
Add `activities-employer-start-date` and `activities-partner-employer-start-date` screens.  Add inputs for each applicants start date of employment and each partners employment start date, if applicable.  Add validation of a date if the user inputs a value.  Create a preparer that iterates over the jobs and partner jobs fields and added values to the pdf.   Added test that verifies the validations, Journeys and Preparers work as expected.  


#### 📷 Design reference
[Figma](https://www.figma.com/design/1oRFY7p6tdGogBJk7Ugle8/IL-CCAP-Families-Application?node-id=13941-15192&m=dev)

#### ✅ Completion tasks

- [ ] Added relevant tests
- [ ] Meets acceptance criteria

analoo commented on src/test/java/org/ilgcc/app/utils/AbstractBasePageTest.java:
```diff
@@ -231,6 +235,7 @@ protected void addParentPartnerJob(String postFix) {
         testPage.enter("partnerEmployerStreetAddress", "123 Partner Employer Address");
         testPage.enter("partnerEmployerZipCode", "6042" + postFix);
         testPage.clickContinue();
+        testPage.clickContinue();
```
what does this do?

enyia21 replied:
clicks continue on the `activities-partner-employer-start-date` screen

analoo replied:
Would you add the name of the screen as a comment?

cram-cfa commented on src/main/java/org/ilgcc/app/submission/actions/ValidateEmployerStartDate.java:
```diff
@@ -0,0 +1,47 @@
+  
+  @Override
+  public Map<String, List<String>> runValidation(FormSubmission formSubmission, Submission submission) {
+    log.info("Running ValidateJobStartDate");
```
This should be `ValidateEmployerStartDate` so at least finding the class from the logging will be obvious

enyia21 replied:
What does this mean?

cram-cfa commented on src/main/java/org/ilgcc/app/submission/actions/ValidateEmployerStartDate.java:
```diff
@@ -0,0 +1,47 @@
+    String employerStartDate = DateUtilities.getFormattedDateFromMonthDateYearInputs("activitiesJobStart", inputData);
+
+    if (employerStartDate.replace("/", "").isBlank()) {
+      return errorMessages;
```
Should this return something more than the empty HashMap? (ie, having the error that the date is empty?)

analoo replied:
blank dates are allowed in this case.

cram-cfa replied:
Oh, I get what it's doing now! I thought it was just forgetting to add the correct error message :)

cram-cfa commented on src/main/java/org/ilgcc/app/submission/actions/ValidatePartnerEmployerStartDate.java:
```diff
@@ -0,0 +1,47 @@
+
+  @Override
+  public Map<String, List<String>> runValidation(FormSubmission formSubmission, Submission submission) {
+    log.info("Running ValidateJobStartDate");
```
Same here, should be `ValidatePartnerEmployerStartDate `

enyia21 replied:
I see what you're saying

cram-cfa reviewed (changes requested):
Is there a way to combine `ValidatePartnerEmployerStartDate` and `ValidateEmployerStartDate` into some shared code? The logic looks the same, just different input fields and error message keys...

analoo commented:
Agree with Marc's comments and will pre-emptively approved now since I tested the branch and it works as expected:
- empty dates are allowed but filled in dates are validated for both parent and parent partner.
- the dates correctly printed to the PDF

analoo commented on src/test/java/org/ilgcc/app/utils/SubmissionTestBuilder.java:
```diff
@@ -248,7 +248,32 @@ public SubmissionTestBuilder withJob(String subflow, String companyName, String
         submission.getInputData().put(subflow, jobs);
         return this;
     }
+    public SubmissionTestBuilder withJob(String subflow, HashMap<String, Object> fields) {
```
I think the preparer can be cleaner. This his not a blocking comment by any means.

Rather than adding a new class called withJob with a different signature, I would try to incorporate the activitiesStart date into the current test builders.

enyia21 commented:
@cram-cfa and @analoo 
Do these refactors meet your expectations?
I've:
- [ ] combined the ValidateEmployerStartDate into one validation and refactored the test to take this change into account.
- [ ] I've refactored the EmployerStartDate Preparer to be cleaner looking.   
- [ ] I've removed the ValidatePartnerEmployerStartDate validation and its test.

cram-cfa reviewed (approved):
LGTM!
